src/benchmarks/micro/coreclr/Devirtualization/GuardedTwoClassVirtual.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using BenchmarkDotNet.Attributes;
+using MicroBenchmarks;
+
+// Performance test for virtual call dispatch with two
+// possible target classes mixed in varying proportions.
+
+namespace GuardedDevirtualizationTwoClass
+{
+
+public class B
+{
+   public virtual int F() => 33;
+}
+
+public class D : B
+{
+    public override int F() => 44;
+}
+
+public class TestInput
+{
+    public const int N = 1000;
+
+    public TestInput(double pB)
+    {
+        _pB = pB;
+        b = GetArray();
+    }
+
+    static Random r = new Random(42);
+
+    double _pB;
+    B[] b;
+
+    public B[] Array => b;
+    public override string ToString() => $"pB = {_pB:F2}";
+
+    B[] GetArray()
+    {
+        B[] result = new B[N];
+        for (int i = 0; i < N; i++)
+        {
+            double p = r.NextDouble();
+            if (p > _pB)
+            {
+                result[i] = new D();
+            }
+            else
+            {
+                result[i] = new B();
+            }
+        }
+        return result;
+    }
+}
+
+public class Virtual
+{
+    [BenchmarkCategory(Categories.CoreCLR, Categories.Virtual)]
+    [Benchmark(OperationsPerInvoke=TestInput.N)]
+    [ArgumentsSource(nameof(GetInput))]
+    public static long Call(TestInput testInput)
+    {
+        long sum = 0;
+        B[] input = testInput.Array;
+        for (int i = 0; i < input.Length; i++)
+        {
+            sum += input[i].F();
+        }
+        return sum;
+    }
+
+    static int S = 10;
+    static double delta = 1.0 / (double) S;
+
+    public IEnumerable<TestInput> GetInput()
```

Comment:
when I was porting the old CoreFX benchmarks to the perf repo I found few benchmarks with hundreds of test cases (600, 400, 200). It was clearly a human error, but responsible for half of the execution time of the CoreFX suite ;)

To prevent users from adding too many test cases I have added a `TooManyTestCasesValidator` which does not allow for mor than 16 test cases per benchmark.

If you really want to test more than 16 test cases locally you need to comment this line:

https://github.com/dotnet/performance/blob/2a7df6527d2deec682ae3acae18b39bd8901ea9a/src/benchmarks/micro/Program.cs#L38

Reply:
I'll cut down the number of sample points for now, but it would be nice to be able to bypass this somehow.

Or I could split it up as multiple tests.